scipy/stats/_morestats.py

```diff
@@ -1309,27 +1321,33 @@ def _all(x):
         message = ("The `optimizer` argument of `boxcox_normmax` must return "
                    "an object containing the optimal `lmbda` in attribute `x`.")
         raise ValueError(message)
-    else:
-        # Test if the optimal lambda causes overflow
+    elif ymax is None or not np.isinf(ymax):  # adjust the final lambda
         x = np.asarray(x)
-        max_x = np.max(x, axis=0)
-        istransinf = np.isinf(special.boxcox(max_x, res))
-        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
-        if np.any(istransinf):
-            warnings.warn(
+        xmax = np.max(x, axis=0)
+        ymax_res = special.boxcox(xmax, res)
+
+        if ymax is None:
+            # The default `ymax` is the maximum value of the input dtype
+            dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
+            ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
+            end_msg = f"cause overflow in {dtype}."
+        else:
+            end_msg = "exceed specified `ymax`."
+
+        mask = ymax_res > ymax
+        if np.any(mask):
+            message = (
                 f"The optimal lambda is {res}, but the returned lambda is "
                 f"the constrained optimum to ensure that the maximum of the "
-                f"transformed data does not cause overflow in {dtype}.",
-                stacklevel=2
+                f"transformed data does not " + end_msg
             )
+            warnings.warn(message, stacklevel=2)
 
             # Return the constrained lambda to ensure the transformation
-            # does not cause overflow
-            ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
-            constrained_res = _boxcox_inv_lmbda(max_x, ymax)
-
+            # does not cause overflow or exceed specified `ymax``
+            constrained_res = _boxcox_inv_lmbda(xmax, ymax)
```

Comment:
I think `_boxcox_inv_lmbda` needs to pass `k=-1` into `lambertw`. I had this originally and changed it by mistake, I think.

<img width="711" alt="image" src="https://github.com/scipy/scipy/assets/6570539/950184bd-2939-4914-9d06-a0a98fddc258">